pkg/database/user_report.go

```diff
@@ -110,13 +120,74 @@ func (db *Database) FindUserReport(tx *gorm.DB, phoneNumber string) (*UserReport
 
 // DeleteUserReport removes a specific phone number from the user report
 // de-duplication table.
-func (db *Database) DeleteUserReport(phoneNumber string) error {
+// The actor may be nil on this operation, if it's the result of a
+// natural delete (SMS failure).
+// Otherwise this is an administrative acction.
+func (db *Database) DeleteUserReport(phoneNumber string, actor Auditable) error {
 	hmacedCodes, err := db.generatePhoneNumberHMACs(phoneNumber)
 	if err != nil {
 		return fmt.Errorf("failed to create hmac: %w", err)
 	}
 
-	return db.db.Where("phone_hash IN (?)", hmacedCodes).Delete(&UserReport{}).Error
+	return db.db.Transaction(func(tx *gorm.DB) error {
+		var ur UserReport
+		if err := tx.
+			Set("gorm:query_option", "FOR UPDATE").
+			Model(&UserReport{}).
+			Where("phone_hash IN (?)", hmacedCodes).
+			First(&ur).
+			Error; err != nil {
+			if IsNotFound(err) {
+				// Log an audit record on attempted purges as well.
+				if actor != nil {
+					// Created to furfill the audit requirements.
+					ur = UserReport{
+						ID:          0,
+						PhoneHash:   hmacedCodes[0],
+						CodeClaimed: false,
+					}
+					audit := BuildAuditEntry(actor, "attempted to purge user report phone", &ur, 0)
+					if err := tx.Save(audit).Error; err != nil {
+						return fmt.Errorf("failed to save audits: %w", err)
+					}
+				}
+			}
+			return nil
+		}
+
+		vc := &VerificationCode{}
+		if err := tx.
+			Set("gorm:query_option", "FOR UPDATE").
+			Model(&VerificationCode{}).
+			Where("user_report_id = ?", ur.ID).
+			First(vc).
+			Error; err != nil {
+			if IsNotFound(err) {
+				vc = nil
+			} else {
+				return err
+			}
+		}
+		if vc != nil && !vc.IsExpired() {
```

Comment:
It might be better to call `realm.ExpireCode` here instead of share the logic.

Reply:
it's a different entry point - UUID vs user_report_id, that a substantial refactor. 

I think we want it this way w/ the audit record being associated w/ the user report and not w/ the verification code